src/lazyoperations.jl

```diff
@@ -41,24 +55,272 @@ function det(K::Kron{<:Any, 2})
             # The Kronecker Product of rectangular matrices, if it is square, will
             # have determinant zero. This can be shown by using the fact that
             # rank(A ⊗ B) = rank(A)rank(B) and showing that this is strictly less
-            # than the number of rows in the resulting Kronecker matrix. Hence,
+            # than the number of rows/cols in the resulting Kronecker matrix. Hence,
             # since A ⊗ B does not have full rank, its determinant must be zero.
-            return zero(d)
+            return zero(T)
         end
     end
     return d
 end
 
+
+function logabsdet(K::Kron{T, 2}) where T<:Number
+    s = checksquare(K)
+    d = zero(T)
+    sgn = one(T)
+
+    for A in K.args
+        if issquare(A)
+            ldA, sgnA = logabsdet(A)
+            if isinf(ldA)
+                return ldA, sgnA
+            end
+            p = (s ÷ size(A, 1))
+            d += p * ldA
+            sgn *= sgnA ^ p
+        else
+            # see definition of det above for explanation
+            return float(T)(-Inf), zero(T)
+        end
+    end
+    return d, sgn
+end
+
+
 function tr(K::Kron{<:Any, 2})
-    (size(K, 1) == size(K, 2)) || throw(DimensionMismatch("matrix is not square: dimensions are $(size(K))"))
-    if all(A -> (size(A, 1) == size(A, 2)), K.args)  # check if all component matrices are square
+    checksquare(K)
+    if all(issquare, K.args)
         return prod(tr.(K.args))
     else
         return sum(diag(K))
     end
 end
 
 
+function diag(K::Kron{<:Any, 2})
+    if all(issquare, K.args)
+        return kron(diag.(K.args)...)
+    else
+        d = similar(K.args[1], minimum(size(K)))
+        @inbounds for i in 1:length(d)
+            d[i] = K[i, i]
+        end
+        return d
+    end
+end
+
+
+_compatible_sizes((A, B)) = (size(A, 2) == size(B, 1))
+
+
+# Implements the mixed-produce property for the Kronecker product and matrix
+# multiplication
+function materialize(M::Applied{
+    MulAddStyle, typeof(*),
+    NTuple{2, Kron{T,2,NTuple{2, MT}}}
+}) where {T, MT<:AbstractMatrix}
+    A, B = M.args
+    # Keeping it simple for now, but could potentially make this "alignment"-check
+    # more precise. For example, if one factor of A has 6 columns and it's aligned
+    # with two factors of B with 2 and 3 rows, then we can still use the
+    # mixed-product property after explicitly computing the Kronecker product of
+    # the two factors of B (or maybe just deferring to the shuffle algorithm)
+    if (length(A.args) == length(B.args)) && all(_compatible_sizes, zip(A.args, B.args))
+        factors = [(A_i * B_i) for (A_i, B_i) in zip(A.args, B.args)]
+        return Kron(factors...)
+    else
+        algo_type = shuffle_algorithm_type(MT)
+        return shuffle_algorithm(algo_type, A, B, eltype(M))
+    end
+end
```

Comment:
The following would be better to embrace the `MulAdd` type from ArrayLayouts.jl., to support in-place operations. For the case `(length(A.args) == length(B.args)) && all(_compatible_sizes, zip(A.args, B.args))` this seems fairly straightforward, something like:
```julia
similar(M::MulAdd{ApplyLayout{typeof(kron)}, ApplyLayout{typeof(kron)}, ::Type{T}) where T = Kron(similar.(MulAdd.(arguments(M.A), arguments(M.B)), T)...)

function materialize!(M:MulAdd{ApplyLayout{typeof(kron)}, ApplyLayout{typeof(kron)},ApplyLayout{typeof(kron)}})
   materialize!.(MulAdd.(M.α, arguments(M.A), arguments(M.B), M.β, arguments(M.C))
   M.C
end
```

Though I don't know how hard the other case would be... so happy to merge as is.

Technically you should be making a new `ApplyStyle` and not overload with `MulAddStyle`.

Reply:
I don't understand the different `ApplyStyle`s or the types in ArrayLayouts.jl very well, so I'm not sure I'll be able to figure this out without help.